daemon/util/Thread.cpp

```diff
@@ -41,29 +42,22 @@ void Thread::Start()
 {
 	debug("Starting Thread");
 
-	m_running = true;
+	std::promise<void> ready;
+	std::future<void> futureReady = ready.get_future();
 
-	// NOTE: we must guarantee, that in a time we set m_running
-	// to value returned from pthread_create, the thread-object still exists.
-	// This is not obvious!
-	// pthread_create could wait long enough before returning result
-	// back to allow the started thread to complete its job and terminate.
-	// We lock mutex m_threadMutex on calling pthread_create; the started thread
-	// then also try to lock the mutex (see thread_handler) and therefore
-	// must wait until we unlock it
-	Guard guard(m_threadMutex);
+	// start the new thead and signal when ready
+	std::thread t([&]{
+		m_running = true;
+		ready.set_value();
+		thread_handler();
+	});
 
-#ifdef WIN32
-	m_threadObj = (HANDLE)_beginthread(Thread::thread_handler, 0, (void*)this);
-	m_running = m_threadObj != 0;
-#else
-	pthread_attr_t m_attr;
-	pthread_attr_init(&m_attr);
-	pthread_attr_setdetachstate(&m_attr, PTHREAD_CREATE_DETACHED);
-	pthread_attr_setinheritsched(&m_attr, PTHREAD_INHERIT_SCHED);
-	m_running = !pthread_create(&m_threadObj, &m_attr, Thread::thread_handler, (void *) this);
-	pthread_attr_destroy(&m_attr);
-#endif
+	// wait for the ready signal
+	futureReady.wait();
+
+	// save the native handler to be able to cancel (Kill) the thread and then detach
+	m_threadObj = t.native_handle();
```

Comment:
At this moment the thread may be already terminated (if it was short-lived). If it was auto-destroying (`m_autoDestroy`) then `this` is already destroyed and we can't access `m_threadObj` anymore.

Would moving this line into thread's lambda function fix the issue?

----

> // save the native handler

A typo: handler -> handle.

Reply:
What is the advantage of using promise/future over a mutex?

Promise/future is a new thing to me and I'm trying to understand how they work and if they are better than manually managed mutexes. 

Promise/future has some kind of thread synchronisation under the hood as well, right? The old code was using one synchronisation (via mutex) to start the thread and increase `m_threadCount`. The new implementation has two synchronisations - one via promise/future and another one via atomic. Generally more synchronisations mean more CPU cache flushes and worse performance.

Reply:
> At this moment the thread may be already terminated (if it was short-lived). If it was auto-destroying (`m_autoDestroy`) then `this` is already destroyed and we can't access `m_threadObj` anymore.

OK, I see what you mean. I think for now I can keep the old mutex but use it only for ensuring this. I wonder if the need for the `m_autoDestroy` can be avoided but I still haven't looked at that part of the code yet.

> What is the advantage of using promise/future over a mutex?

It's higher level and let's the underlying implementation to use the best method to implement it. promise/future are useful to send results and states between threads and are meant to be used only once, which fits nice into this use case.

> The new implementation has two synchronisations - one via promise/future and another one via atomic. 

Using atomic is nice if the thread contention is low as it doesn't stop the threads (it locks but doesn't stop the waiting thread and switch context). In the previous implementation we had a few shared values not protected and using atomic is in my opinion a nice solution as it's more efficient than a mutex (mutexes use atomic operations internally) and thread contention should be very low for this cases.

Reply:
> OK, I see what you mean. I think for now I can keep the old mutex but use it only for ensuring this.

Once this is fixed I can merge the PR.

I've tried this and it worked:

```C++
void Thread::Start()
{
	debug("Starting Thread");

	m_running = true;

	// NOTE: "m_threadMutex" ensures that "t" lives until the very end of the function
	Guard guard (m_threadMutex);

	// start the new thread
	std::thread t([&]{
		{
			// trying to lock "m_threadMutex", this will wait until function "Start()" is completed
			// and "t" is detached.
			Guard guard(m_threadMutex);
		}

		thread_handler();
	});

	// save the native handle to be able to cancel (Kill) the thread and then detach
	m_threadObj = t.native_handle();
	t.detach();
}
```

I've also tried to solve this with promise/future but I needed two pairs of promise/future. The solution with mutex is simpler.

Reply:
Updated